go/client/tree_test.go

```diff
@@ -112,6 +116,242 @@ func TestBuildTree(t *testing.T) {
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a/foo.txt", "bla.txt"},
+			},
+			want: &client.FileTree{
+				Files: map[string][]byte{"bla.txt": []byte("foo")},
+				Dirs: map[string]*client.FileTree{
+					"a": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+				},
+			},
+		},
+		{
+			desc: "file_relative_symlink",
+			input: []*inputPath{
+				{path: "a/foo.txt", fileContents: []byte("foo")},
+				{path: "bla.txt", isSymlink: true, symlinkTarget: "a/foo.txt"},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a/foo.txt", "bla.txt"},
+			},
+			want: &client.FileTree{
+				Files: map[string][]byte{"bla.txt": []byte("foo")},
+				Dirs: map[string]*client.FileTree{
+					"a": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+				},
+			},
+		},
+		{
+			desc: "dir_relative_symlink",
+			input: []*inputPath{
+				{path: "a/foo.txt", fileContents: []byte("foo")},
+				{path: "b", isSymlink: true, symlinkTarget: "a"},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a/foo.txt", "b"},
+			},
+			want: &client.FileTree{
+				Dirs: map[string]*client.FileTree{
+					"a": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+					"b": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+				},
+			},
+		},
+		{
+			desc: "dir_absolute_symlink",
+			input: []*inputPath{
+				{path: "a/foo.txt", fileContents: []byte("foo")},
+				// The absolute symlink target will actually point to the absolute path of a.
+				{path: "b", isSymlink: true, isAbsolute: true, symlinkTarget: "a"},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a/foo.txt", "b"},
+			},
+			want: &client.FileTree{
+				Dirs: map[string]*client.FileTree{
+					"a": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+					"b": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+				},
+			},
+		},
+		{
+			desc: "file_exclusions",
+			input: []*inputPath{
+				{path: "a/foo", fileContents: []byte("foo")},
+				{path: "a/b/bar.txt", fileContents: []byte("bar")},
+				{path: "txt/a", fileContents: []byte("a")},
+				{path: "a/b/baz.txt", fileContents: []byte("baz")},
+				{path: "b/bla", fileContents: []byte("bla")},
+				{path: "bla.txt", fileContents: []byte("bla")},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a", "b", "txt", "bla.txt"},
+				InputExclusions: []*command.InputExclusion{
+					&command.InputExclusion{Regex: `txt$`, Type: command.FileInputType},
+				},
+			},
+			want: &client.FileTree{
+				Dirs: map[string]*client.FileTree{
+					"a":   &client.FileTree{Files: map[string][]byte{"foo": []byte("foo")}},
+					"b":   &client.FileTree{Files: map[string][]byte{"bla": []byte("bla")}},
+					"txt": &client.FileTree{Files: map[string][]byte{"a": []byte("a")}},
+				},
+			},
+		},
+		{
+			desc: "dir_exclusions",
```

Comment:
Can we test directory exclusions for subdirectories?

Reply:
Done.